gateway/src/shard/builder.rs

```diff
@@ -219,8 +222,37 @@ impl ShardBuilder {
     }
 
     /// Consume the builder, constructing a shard.
-    pub fn build(self) -> (Shard, Events) {
-        Shard::new_with_config(self.into_config())
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`ShardStartErrorType::RetrievingGatewayUrl`] error type if
+    /// the gateway URL couldn't be retrieved from the HTTP API.
+    pub async fn build(mut self) -> Result<(Shard, Events), ShardStartError> {
+        if self.gateway_url.is_none() {
+            // By making an authenticated gateway information retrieval request
+            // we're also validating the configured token.
```

Comment:
Could this be split up into its own error so you know which of the issues you ran into.

Though that could also be done in another pr since it is just copy/paste here